app/src/main/java/com/github/sdp/tarjetakuna/database/ChatsRTDB.kt

```diff
@@ -0,0 +1,184 @@
+package com.github.sdp.tarjetakuna.database
+
+import android.util.Log
+import com.github.sdp.tarjetakuna.model.Chat
+import com.github.sdp.tarjetakuna.model.Message
+import com.google.android.gms.tasks.Task
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.DatabaseReference
+import com.google.firebase.database.ValueEventListener
+import com.google.gson.Gson
+import java.util.concurrent.CompletableFuture
+
+/**
+ * This class is used to manage the chats table in the database.
+ */
+class ChatsRTDB(database: Database = FirebaseDB()) {
+
+    private var chatsTable: DatabaseReference
+    private var messagesRTDB: MessagesRTDB
+    private var chatsEventListener: HashMap<String, ValueEventListener> = HashMap()
+
+    companion object {
+        private const val TAG = "ChatsRTDB"
+    }
+
+    init {
+        chatsTable = database.chatsTable() //Firebase.database.reference.child("chats")
+        messagesRTDB = MessagesRTDB(database)
+    }
+
+    /**
+     * Add a chat to the chats table.
+     */
+    fun addChat(chat: DBChat): Task<Void> {
+        return chatsTable.child(chat.uid).setValue(chatToDBFormat(chat))
+    }
+
+    /**
+     * Get a chat from the chats table.
+     */
+    fun getChat(chatUID: String): CompletableFuture<DBChat> {
+        val future = CompletableFuture<DBChat>()
+        chatsTable.child(chatUID).get().addOnSuccessListener {
+            if (it.value == null) {
+                future.completeExceptionally(NoSuchFieldException("chat $chatUID is not in chats table"))
+            } else {
+                future.complete(chatFromDBFormat(it))
+            }
+        }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+        return future
+    }
+
+    /**
+     * Remove a chat from the chats table.
+     */
+    fun removeChat(chatUID: String): Task<Void> {
+        return chatsTable.child(chatUID).removeValue()
+    }
+
+    /**
+     * Add a chat to the database.
+     */
+    fun addChatToDatabase(chat: Chat): ArrayList<Task<Void>> {
+        val tasks = ArrayList<Task<Void>>()
+        val task = addChat(DBChat.toDBChat(chat))
+        tasks.add(task)
+        for (message in chat.messages) {
+            val msgTask = messagesRTDB.addMessageToDatabase(message)
+            tasks.add(msgTask)
+        }
+        return tasks
+    }
+
+    /**
+     * Get a chat from the database.
+     */
+    fun getChatFromDatabase(chatUID: String): CompletableFuture<Chat> {
+        val future = CompletableFuture<Chat>()
+        getChat(chatUID).thenAccept {
+            val chat = DBChat.fromDBChat(it)
+            messagesRTDB.getMessagesFromDatabase(it.messages).thenAccept { it2 ->

```

Comment:
Does this work with the nested thenAccept? I personally had some problems when trying to implement something similar in UserRTDB and had to use thenCompose.

Reply:
yeah it depends what you return but it worked i think, let me double check

Reply:
just made some debugging test and yes it works !